File: json_to_flutter/functions/widgetInfo.dart
import 'dart:math';
import '../models/index.dart';

No attWidgetInfo(No no) {
  if (no.widget.id == "-1") {
    Container newInfo = _newWidgetInfo(no.children);
    no.widget.x = newInfo.x;
    no.widget.y = newInfo.y;
    no.widget.gw = newInfo.gw;
    no.widget.gh = newInfo.gh;
  }
  if (no.father != null) {
    attWidgetInfo(no.father);
  }
  return no;
}

Widget _newWidgetInfo(List<No> nos) {
  double x = -1, y = -1, w = -1, h = -1;
  nos.forEach((f) {
    if (x == -1) {
      x = f.widget.x;
      y = f.widget.y;
      w = x + f.widget.gw;
      h = y + f.widget.gh;
    } else {
      x = min(x, f.widget.x);
      y = min(y, f.widget.y);
      w = max(w, f.widget.x + f.widget.gw);
      h = max(h, f.widget.y + f.widget.gh);
    }
  });
  return new Container(x, y, 0, 0, w - x, h - y);
}

File: src/app/json_to_flutter/functions/widgetInfo.js
function attWidgetInfo(no) {
    if (no.widget.id == "-1") {
        let newInfo = _newWidgetInfo(no.children);
        no.widget.x = newInfo[0];
        no.widget.y = newInfo[1];
        no.widget.w = newInfo[2];
        no.widget.h = newInfo[3];
    }
    if (no.father != null) {
        attWidgetInfo(no.father);
    }
    return no;
}

function _newWidgetInfo(nos) {
    let x = -1, y = -1, w = -1, h = -1;
    nos.forEach(function (f) {
        if (x == -1) {
            x = f.widget.x;
            y = f.widget.y;
            w = x + f.widget.w;
            h = y + f.widget.h;
        } else {
            x = Math.min(x, f.widget.x);
            y = Math.min(y, f.widget.y);
            w = Math.max(w, f.widget.x + f.widget.w);
            h = Math.max(h, f.widget.y + f.widget.h);
        }
    });
    return [x, y, w - x, h - y];
}

module.exports = { attWidgetInfo };